Feature: Add more filtering and favourites to build object/station/waypoint/roadstop windows.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Objects, Stations, Waypoints and Roadstops all support selecting add-on content from NewGRF files.

The original design is that you select from a list of classes (up to 15 could be used!), and then pick a type from a
list.

Obviously 15 wasn't enough, so at some point this was extended, and a search filter was added. This filter can only
search from the names of classes, not the names of types within.

These windows have been developed 'ad-hoc' over the years, and although they mostly share the same layout and
functionality, everything is duplicated and no code is shared between them, so changing anything has to be done at least
4 or 5 times...

Last but not least the Build Object window has some interesting design that means the window expands based on NewGRF content, and this can expand to ridiculous proportions.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Introducing the unified-code build-picker window. This replaces much of the similar code in each of the build windows
with a shared window handler. Helper classes/callbacks are used to determine per-feature specifics, but the general
function is performed by the shared code.

Along with unifying the code behing these windows, this allow news features to be added for all features all at once:

* Additional text search for types. You can now search for within classes for a specific item.
* A new "All" button that 'ignores' the class selection list (but not the class text filter) and displays all types
  together. This makes the type text search even more useful.
* A new "Used" button, which will display only types that have already been placed by the current company. When lots of
  add-on content is loaded, this makes choosing from a more limited subset much easier, as you no longer have to scroll
  through a list of 200+ classes...
* Favourites! Ctrl-click on an item in the list of types to mark it as a favourite. This then appears with other saved
  items under the new "Saved" button. The favourites are saved to a local configuration file (favs.cfg) so are
  remembered after quitting and restarting.

![image](https://github.com/OpenTTD/OpenTTD/assets/639850/526d18c8-a1ab-4de2-9c79-cac16b6ee843)

![image](https://github.com/OpenTTD/OpenTTD/assets/639850/946ca828-8946-4269-8c0d-36a101a79ccb)

And this adds less lines of code than is removed...

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

This is pretty large, but I've built the commits up in a way that can be broken into smaller PRs if required.

Probably some bugs, there's a lot going on!

The way the looping works is kind of a backwards step, but it was intended to be generic. Although all the buildable stuff right now uses NewGRFClass, even within that framework things are not consistent (waypoints, bus vs truck, etc). And there's also the possibility of introducing a build picker for houses, which don't use the NewGRFClass system.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

2TallTyler commented on src/newgrf_class.h:
```diff
@@ -12,21 +12,31 @@
+/* Base for each type of NewGRF spec to be used with NewGRFClass. */
+template <typename Tindex>
+struct NewGRFSpecBase {
+	Tindex class_index; ///< Class index of thie spec, invalid until class is allocated.
```
```suggestion
	Tindex class_index; ///< Class index of this spec, invalid until class is allocated.
```

2TallTyler commented on src/newgrf_roadstop.h:
```diff
@@ -181,4 +181,14 @@ int AllocateSpecToRoadStop(const RoadStopSpec *statspec, BaseStation *st, bool e
+/**
+ * Test if a RoadStopClass is the waypoint class.
+ * @param cls RoadStopClass to test.
+ * @return true if the class the waypoint class.
```
```suggestion
 * @return true if the class is the waypoint class.
 ```

2TallTyler commented on src/newgrf_station.h:
```diff
@@ -181,6 +181,16 @@ using StationClass = NewGRFClass<StationSpec, StationClassID, STAT_CLASS_MAX>;
+/**
+ * Test if a StationClass is the waypoint class.
+ * @param cls StationClass to test.
+ * @return true if the class the waypoint class.
```
```suggestion
 * @return true if the class is the waypoint class.
```

2TallTyler commented on src/picker_gui.h:
```diff
@@ -0,0 +1,176 @@
+	static const uint EDITBOX_MAX_SIZE = 16; ///< The maximum number of characters for the filter edit box.
+
+	bool class_picker;
+	bool type_picker;
```
These confused me when reviewing their implementation, because they are not named like bools — at first I thought you meant to check `if (class_picker != nullptr)`

Maybe `use_class_picker` or `has_class_picker` would make this clearer?

2TallTyler commented on src/picker_gui.cpp:
```diff
@@ -0,0 +1,463 @@
+			if (FillDrawPixelInfo(&tmp_dpi, ir)) {
+				AutoRestoreBackup dpi_backup(_cur_dpi, &tmp_dpi);
+				int x = (ir.Width()  - ScaleSpriteTrad(64)) / 2 + ScaleSpriteTrad(31);
+				int y = (ir.Height() + ScaleSpriteTrad(58)) / 2 - ScaleSpriteTrad(31);
```
Are there constants that could be used instead of these magic numbers? ("No" is a valid answer. 🙂)

PeterN replied:
![image](https://github.com/OpenTTD/OpenTTD/assets/639850/a865202b-0aee-43e7-8bc3-b21ef649f528)
I'm going with "should be but out of scope"

2TallTyler commented on src/road_gui.cpp:
```diff
@@ -1093,144 +1093,137 @@ static void ShowRoadDepotPicker(Window *parent)
+	{
+		const auto *spec = this->GetSpec(cls_id, id);
+		if (spec == nullptr) {
+			StationPickerDrawSprite(x, y, roadstoptype == ROADSTOP_BUS ? STATION_BUS : STATION_TRUCK, INVALID_RAILTYPE, _cur_roadtype, _roadstop_gui.orientation);
```
Looks like you copied and pasted this, but should this be `INVALID_ROADTYPE`?

```suggestion
			StationPickerDrawSprite(x, y, roadstoptype == ROADSTOP_BUS ? STATION_BUS : STATION_TRUCK, INVALID_ROADTYPE, _cur_roadtype, _roadstop_gui.orientation);
```

PeterN replied:
Nope, this is rabbit hole where StationPickerDrawSprite is used for both rail and road stations.

2TallTyler commented on src/picker_gui.cpp:
```diff
@@ -352,9 +363,23 @@ void PickerWindow::BuildPickerTypeList()
+		for (int class_index : this->classes) {
+			int count = this->callbacks.GetTypeCount(class_index);
+			for (int i = 0; i < count; i++) {
+				if (this->callbacks.GetTypeName(class_index, i) == STR_NULL) continue;
```
Does this ignore NewGRFs that use an empty string for the type name? (If that's even possible) Or would that be `STR_EMPTY`?

PeterN replied:
Changed it to `INVALID_STRING_ID` instead, which I don't think is possible for NewGRFs to use.

2TallTyler reviewed (commented):
This is a fantastic quality of life feature. Code looks generally good to me, and while the GUI diffs are hard to read I know you'll fix any bugs that slip past reviews. 😉

The saved feature is fantastic, although it doesn't work in emscription so I can't test the following questions without checking out and building locally. Am I reading the code correctly that the saved list is independent of what GRFs you have loaded, and the load order? So I can save my favorite stations, for example, and if I have the GRF loaded they'll show up as saved? And if the GRF is not loaded, nothing will happen? (And the saved list won't be wiped?)

I did find one bug in emscription, which is that the selected item in the picker is not updated when I filter by used or saved -- the orientation preview and name string show whatever was selected in the previous unfiltered view.

Sometimes an object in the filtered list is selected... (the wrong one)

![selected](https://github.com/OpenTTD/OpenTTD/assets/55058389/b27ebd21-b18e-4222-9f96-e13eead1dfc7)

...and sometimes not.

![unselected](https://github.com/OpenTTD/OpenTTD/assets/55058389/38b0e247-eb7d-40ec-baba-b3426ede4dbd)

2TallTyler commented on src/widgets/picker_widget.h:
```diff
@@ -0,0 +1,35 @@
+	WID_PW_CLASS_FILTER, ///< Editbox filter.
+	WID_PW_CLASS_LIST, ///< List of classes.
+	WID_PW_CLASS_SCROLL, ///< Scrollbar for list of classes.
+	WID_PW_CLASS_RESIZE, ///< Class resize handle, if standalone window.
```
Is this standalone window mode part of this PR, or leftover from a different iteration of the patch? (I remember seeing screenshots where the window could be split up...)

2TallTyler commented on src/widgets/picker_widget.h:
```diff
@@ -0,0 +1,34 @@
+	WID_PW_TYPE_ITEM, ///< A single item.
+	WID_PW_TYPE_SCROLL, ///< Scrollbar for the matrix.
+	WID_PW_TYPE_NAME, ///< Name of selected item.
+	WID_PW_TYPE_RESIZE, ///< Type resize handle, if standalone window.
```
Missed a `standalone window` reference. 🙂

rubidium42 commented on src/object_gui.cpp:
```diff
@@ -34,208 +34,129 @@
-		this->object_classes.Filter(this->string_filter);
-		this->object_classes.RebuildDone();
-		this->object_classes.Sort();
+	static ObjectPickerCallbacks instance;
+};
+/* static */ ObjectPickerCallbacks ObjectPickerCallbacks::instance;
```
Maybe `static inline`?

PeterN replied:
Not possible as it's an instance of itself.

rubidium42 commented on src/object_gui.cpp:
```diff
@@ -246,41 +167,14 @@ class BuildObjectWindow : public Window {
+				const ObjectClass *objclass = ObjectClass::Get(_object_gui.sel_class);
+				const ObjectSpec *spec = objclass->GetSpec(_object_gui.sel_type);
+				size.width  = ScaleGUITrad(64) + WidgetDimensions::scaled.fullbevel.Horizontal();
+				size.height = ScaleGUITrad(58) + WidgetDimensions::scaled.fullbevel.Vertical();
```
The 64 and 56 are used later in this class again. Would it make sense to have these as the same named constant?

PeterN replied:
They're actually used all over the place. I think resolving that is a job for a separate PR.

rubidium42 commented on src/rail_gui.cpp:
```diff
@@ -929,65 +929,109 @@ static void HandleStationPlacement(TileIndex start, TileIndex end)
-		this->vscroll->ScrollTowards(pos);
 	}
 
+	static StationPickerCallbacks instance;
+};
+/* static */ StationPickerCallbacks StationPickerCallbacks::instance;
```
Maybe `static inline`?

rubidium42 commented on src/rail_gui.cpp:
```diff
@@ -1235,43 +1154,8 @@ struct BuildRailStationWindow : public PickerWindowBase {
 			case WID_BRAS_PLATFORM_DIR_Y:
-			case WID_BRAS_IMAGE:
 				size.width  = ScaleGUITrad(64) + WidgetDimensions::scaled.fullbevel.Horizontal();
 				size.height = ScaleGUITrad(58) + WidgetDimensions::scaled.fullbevel.Vertical();
```
I've seen these magic numbers for the object GUI as well, might it be a shared named constant?

rubidium42 commented on src/rail_gui.cpp:
```diff
@@ -2018,232 +1765,115 @@ static void ShowBuildTrainDepotPicker(Window *parent)
-		if (!gui_scope) return;
-		this->list.ForceRebuild();
-	}
+	static WaypointPickerCallbacks instance;
+};
+/* static */ WaypointPickerCallbacks WaypointPickerCallbacks::instance;
```
`static inline`?

rubidium42 commented on src/station_cmd.cpp:
```diff
@@ -1335,8 +1335,10 @@ CommandCost CmdBuildRailStation(DoCommandFlag flags, TileIndex tile_org, RailTyp
 	/* Check if the given station class is valid */
-	if ((uint)spec_class >= StationClass::GetClassCount() || spec_class == STAT_CLASS_WAYP) return CMD_ERROR;
-	if (spec_index >= StationClass::Get(spec_class)->GetSpecCount()) return CMD_ERROR;
+	if ((uint)spec_class >= StationClass::GetClassCount()) return CMD_ERROR;
```
```suggestion
	if (static_cast<uint>(spec_class) >= StationClass::GetClassCount()) return CMD_ERROR;
```

rubidium42 commented on src/station_cmd.cpp:
```diff
@@ -1946,10 +1948,12 @@ CommandCost CmdBuildRoadStop(DoCommandFlag flags, TileIndex tile, uint8_t width,
 	/* Check if the given station class is valid */
-	if ((uint)spec_class >= RoadStopClass::GetClassCount() || spec_class == ROADSTOP_CLASS_WAYP) return CMD_ERROR;
-	if (spec_index >= RoadStopClass::Get(spec_class)->GetSpecCount()) return CMD_ERROR;
+	if ((uint)spec_class >= RoadStopClass::GetClassCount()) return CMD_ERROR;
```
```suggestion
	if (static_cast<uint>(spec_class) >= RoadStopClass::GetClassCount()) return CMD_ERROR;
```

rubidium42 commented on src/waypoint_cmd.cpp:
```diff
@@ -177,8 +177,10 @@ CommandCost CmdBuildRailWaypoint(DoCommandFlag flags, TileIndex start_tile, Axis
 	/* Check if the given station class is valid */
-	if (spec_class != STAT_CLASS_WAYP) return CMD_ERROR;
-	if (spec_index >= StationClass::Get(spec_class)->GetSpecCount()) return CMD_ERROR;
+	if ((uint)spec_class >= StationClass::GetClassCount()) return CMD_ERROR;
```
```suggestion
	if (static_cast<uint>(spec_class) >= StationClass::GetClassCount()) return CMD_ERROR;
```

rubidium42 commented on src/picker_gui.cpp:
```diff
@@ -0,0 +1,644 @@
+		case WID_PW_TYPE_ITEM:
+			size.width  = ScaleGUITrad(64) + WidgetDimensions::scaled.fullbevel.Horizontal();
+			size.height = ScaleGUITrad(58) + WidgetDimensions::scaled.fullbevel.Vertical();
+			break;
```
I've seen these magic numbers in the object and station UIs. Are they the same?

rubidium42 commented on src/road_gui.cpp:
```diff
@@ -1239,86 +1251,12 @@ struct BuildRoadStationWindow : public PickerWindowBase {
 		this->PickerWindowBase::Close();
```
Might this need to be?
```suggestion
		this->PickerWindow::Close();
```

PeterN committed:
Codechange: Add NewGRFSpecsBase class to hold class/index information.

Standardises how the class index is stored in the spec, instead of relying ot the Spec structs having the same members.

This allows retrieving class_index and index without searching or using pointer arithmetic.

'cls_id' is renamed to 'class_index' to make it clearer that it is an index rather than the multichar label of the class.

PeterN committed:
Codechange: Add functions to test if a station/roadstop class is a waypoint.

This is now checked by class label instead of by index.

PeterN committed:
Codechange: Add station/roadstop class labels for DFLT/WAYP.

This avoids repeating the same multichar literal value.

PeterN committed:
Feature: Add "All" filter to build-picker show types from all classes.

Toggling the "All" filter causes the class selection to be ignored, so
that items from all classes can be displayed together. The class text
filter is still applied.

This makes it easier to search amongst types for a feature.

PeterN committed:
Change: Unify station/waypoint/roadstop/object build-picker window code.

These windows now share a common code base for choosing and display class and types.

An additional text filter is added to search types by name instead of just classes.

PeterN committed:
Feature: Ctrl-click to toggle favourites in build-pickers.

This allows ctrl-click on a type in a build-picker window to remember it
as a favourite. An new filter button to show only favourites makes it
simpler to use these types.

Favourite types are saved locally in favs.cfg, so are remembered between
games.

PeterN committed:
Feature: New filter to show only used types in build-pickers.

This filters the build-picker type lists to only show types that have
already been placed in the current game, making it simpler to get to
build matching features.

rubidium42 commented on src/picker_gui.cpp:
```diff
@@ -0,0 +1,644 @@
+	this->InvalidateData(PFI_CLASS | PFI_TYPE | PFI_POSITION | PFI_VALIDATE);
+}
+
+void PickerWindow::Close([[maybe_unused]] int data)
```
I'd wager a bet that `data` is used.
```suggestion
void PickerWindow::Close(int data)
```

rubidium42 commented on src/picker_gui.cpp:
```diff
@@ -0,0 +1,644 @@
+	this->PickerWindowBase::Close(data);
+}
+
+void PickerWindow::UpdateWidgetSize(WidgetID widget, Dimension &size, [[maybe_unused]] const Dimension &padding, [[maybe_unused]] Dimension &fill, [[maybe_unused]] Dimension &resize)
```
`padding` and `resize` are used.
```suggestion
void PickerWindow::UpdateWidgetSize(WidgetID widget, Dimension &size, const Dimension &padding, [[maybe_unused]] Dimension &fill, Dimension &resize)
```

rubidium42 commented on src/picker_gui.cpp:
```diff
@@ -0,0 +1,644 @@
+	}
+}
+
+void PickerWindow::OnClick([[maybe_unused]] Point pt, WidgetID widget, [[maybe_unused]] int click_count)
```
`pt` is used.
```suggestion
void PickerWindow::OnClick(Point pt, WidgetID widget, [[maybe_unused]] int click_count)
```

rubidium42 commented on src/picker_gui.cpp:
```diff
@@ -0,0 +1,644 @@
+	}
+}
+
+void PickerWindow::OnInvalidateData([[maybe_unused]] int data, [[maybe_unused]] bool gui_scope)
```
```suggestion
void PickerWindow::OnInvalidateData(int data, bool gui_scope)
```

rubidium42 commented on src/object_gui.cpp:
```diff
@@ -34,208 +34,129 @@
+		this->PickerWindow::OnInit();
 	}
 
 	void UpdateWidgetSize(WidgetID widget, Dimension &size, [[maybe_unused]] const Dimension &padding, [[maybe_unused]] Dimension &fill, [[maybe_unused]] Dimension &resize) override
```
```suggestion
	void UpdateWidgetSize(WidgetID widget, Dimension &size, const Dimension &padding, Dimension &fill, Dimension &resize) override
```

rubidium42 commented on src/object_gui.cpp:
```diff
@@ -407,131 +247,80 @@ class BuildObjectWindow : public Window {
 		this->SetDirty();
 	}
 
 	void OnInvalidateData([[maybe_unused]] int data = 0, [[maybe_unused]] bool gui_scope = true) override
```
```suggestion
	void OnInvalidateData(int data = 0, bool gui_scope = true) override
```

rubidium42 commented on src/object_gui.cpp:
```diff
@@ -407,131 +247,80 @@ class BuildObjectWindow : public Window {
+		}
 	}
 
 	void OnClick([[maybe_unused]] Point pt, WidgetID widget, [[maybe_unused]] int click_count) override
```
```suggestion
	void OnClick(Point pt, WidgetID widget, int click_count) override
```

rubidium42 commented on src/picker_gui.h:
```diff
@@ -0,0 +1,224 @@
+	explicit PickerCallbacks(const std::string &ini_group);
+	virtual ~PickerCallbacks();
+
+	virtual void Close([[maybe_unused]] int data) { }
```
I do not think it makes much sense annotating that the parameters are unused in here. Or is it required that both the declaration and definition are in sync? If so, then I've missed a lot. If not, then I've also not added a comment to each of the locations where it should be removed.
```suggestion
	virtual void Close(int data) { }
```

rubidium42 commented on src/rail_gui.cpp:
```diff
@@ -1063,138 +1090,35 @@ struct BuildRailStationWindow : public PickerWindowBase {
+		this->PickerWindow::Close();
 	}
 
 	void OnInvalidateData([[maybe_unused]] int data = 0, [[maybe_unused]] bool gui_scope = true) override
```
```suggestion
	void OnInvalidateData(int data = 0, bool gui_scope = true) override
```

rubidium42 commented on src/rail_gui.cpp:
```diff
@@ -1235,54 +1159,18 @@ struct BuildRailStationWindow : public PickerWindowBase {
 	void UpdateWidgetSize(WidgetID widget, Dimension &size, [[maybe_unused]] const Dimension &padding, [[maybe_unused]] Dimension &fill, [[maybe_unused]] Dimension &resize) override
```
```suggestion
	void UpdateWidgetSize(WidgetID widget, Dimension &size, const Dimension &padding, Dimension &fill, Dimension &resize) override
```

rubidium42 commented on src/road_gui.cpp:
```diff
@@ -1093,232 +1094,169 @@ static void ShowRoadDepotPicker(Window *parent)
+		this->PickerWindow::Close();
 	}
 
 	void OnInvalidateData([[maybe_unused]] int data = 0, [[maybe_unused]] bool gui_scope = true) override
```
```suggestion
	void OnInvalidateData(int data = 0, bool gui_scope = true) override
```

rubidium42 commented on src/road_gui.cpp:
```diff
@@ -1353,52 +1291,23 @@ struct BuildRoadStationWindow : public PickerWindowBase {
 	void UpdateWidgetSize(WidgetID widget, Dimension &size, [[maybe_unused]] const Dimension &padding, [[maybe_unused]] Dimension &fill, [[maybe_unused]] Dimension &resize) override
```
```suggestion
	void UpdateWidgetSize(WidgetID widget, Dimension &size, const Dimension &padding, Dimension &fill, Dimension &resize) override
```
